sympy/tensor/array/sp_utils.py

```diff
@@ -0,0 +1,123 @@
+from sympy.core.sympify import _sympify
+from sympy import Tuple, Basic
+from sympy.tensor.array.ndim_array import NDimArray
+from sympy.tensor.array.sparse_ndim_array import SparseNDimArray
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
+from sympy.core.compatibility import Iterable
+
+import functools
+
+# Row-based linked list sparse matrix(LIL)
+class lil(Basic, NDimArray):
+    """
+    Create a sparse array with Row-based linked list sparse matrix(LIL) format.
+    This data structure is efficient in incremental construction of sparse arrays.
+
+    Note
+    ====
+
+    This is an experimental implementation of algorithm, the compatibility with
+    existing format needs to be verified. Only some basic operations are implemented.
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array.sp_utils import lil
+    >>> a = lil([[0, 1, 0], [2, 0, 0]])
+    >>> a._data
+    [[1], [2]]
+    >>> a._rows
+    [[1], [0]]
+    >>> a[0, 1]
+    1
+    >>> a[0, 1] = 3
+    >>> a.tolist()
+    [[0, 3, 0], [2, 0, 0]]
+    """
+    def __new__(cls, iterable=None, shape=None, **kwargs):
+        shape, flat_list = cls._handle_ndarray_creation_inputs(iterable, shape, **kwargs)
+        shape = Tuple(*map(_sympify, shape))
+        cls._check_special_bounds(flat_list, shape)
+        loop_size = functools.reduce(lambda x,y: x*y, shape) if shape else 0
+
+        # Ideally we should use the Array module for this empty initialization, but
+        # it is not yet supported in SymPy
+        data = cls._empty(shape)
+        rows = cls._empty(shape)
+
+        self = Basic.__new__(cls, data, rows, shape, **kwargs)
+        self._shape = shape
+        self._rank = len(shape)
+        self._loop_size = loop_size
+
+        new_dict = {}
+        if isinstance(iterable, Iterable):
+            iterable = SparseNDimArray(iterable)
+
+        if isinstance(iterable, SparseNDimArray):
+            new_dict = iterable._sparse_array
+        # if it is a dense array, it would be cast to a default sparse array
```

Comment:
The idea comes from SciPy where all sparse array are cast to a CSR format before converting to other format.